HIRS_Utils/src/main/java/hirs/tpm/eventlog/uefi/UefiDevicePath.java

```diff
@@ -0,0 +1,441 @@
+    System.arraycopy(path, UefiConstants.OFFSET_20
+            + offset, vendorData, 0, subTypeLength - UefiConstants.SIZE_16);
+    tmpType += " : Vendor Data = " + HexUtils.byteArrayToHexString(vendorData);
+  } else {
+    tmpType += " : No Vendor Data pesent";
+  }
+    subType = tmpType;
+    return tmpType;
+}
+
+/**
+ * Returns nvm device info.
+ * UEFI Specification, Version 2.8.
+ * Name space Identifier (NSID) and IEEE Extended Unique Identifier (EUI-64):
+ * See Links to UEFI Related Documents
+ * (http://uefi.org/uefi under the headings NVM Express Specification.
+ * @param path
+ * @param offset
+ * @return NVM device info.
+ */
+private String nvmSubType(final byte[] path, final int offset) {
+  String tmpType = "NVM Express Namespace = ";
+  byte[] lengthBytes = new byte[UefiConstants.SIZE_2];
+  System.arraycopy(path, UefiConstants.OFFSET_2 + offset, lengthBytes, 0, UefiConstants.SIZE_2);
+  int subTypeLength = HexUtils.leReverseInt(lengthBytes);
+  byte[] nvmData = new byte[subTypeLength];
+  System.arraycopy(path, UefiConstants.OFFSET_4 + offset, nvmData, 0, subTypeLength);
+  tmpType += HexUtils.byteArrayToHexString(nvmData);
+  subType = tmpType;
+  return tmpType;
+}
+
+/**
+ * BIOS Device Type definition.
+ * From Appendix A of the BIOS Boot Specification.
+ * Only process the Device type.
+ * Status bootHandler pointer, and description String pointer are ignored.
+ * @param path byte array holding the device path.
+ * @return String that represents the UEFI defined BIOS Device Type.
+ */
+private String biosDevicePath(final byte[] path, final int offset) {
+  String devPath = "Legacy BIOS : Type = ";
+  byte devPathType = path[offset + 1];
+  Byte pathType = Byte.valueOf(devPathType);
+  switch (pathType.intValue()) {
+    case UefiConstants.DEVPATH_BIOS_RESERVED: devPath += "Reserved"; break;
+    case UefiConstants.DEVPATH_BIOS_FLOPPY:   devPath += "Floppy"; break;
+    case UefiConstants.DEVPATH_BIOS_HD:       devPath += "Hard Disk"; break;
+    case UefiConstants.DEVPATH_BIOS_CD:       devPath += "CD-ROM"; break;
+    case UefiConstants.DEVPATH_BIOS_PCM:      devPath += "PCMCIA"; break;
+    case UefiConstants.DEVPATH_BIOS_USB:      devPath += "USB"; break;
+    case UefiConstants.DEVPATH_BIOS_EN:       devPath += "Embedded network"; break;
+    case UefiConstants.DEVPATH_BIOS_BEV:      devPath +=
+                                 "Bootstrap Entry Vector (BEV) from an Option ROM";
+    break;
+  default: devPath += "Reserved";
+  break;
+  }
+  subType = devPath;
+  return devPath;
+}
+
+/**
+ * Returns PIWG firmware volume info.
+ * UEFI Specification, Version 2.8.
+ * PIWG Firmware File Section 10.3.5.6:
+ * Contents are defined in the UEFI PI Specification.
+ * @param path
+ * @param offset
+ * @return String that represents the PIWG Firmware Volume Path
+  */
+private String piwgFirmVolFile(final byte[] path, final int offset) {
+  String fWPath = "PIWG Firmware File ";
```

Comment:
Same as before and all instances of subType in these private methods.  If the only purpose is to assign directly to subType, then don't create an inner variable.

Reply:
OK